### ISSUE TITLE
Potential problem patching VSphereMachine.Status leads to redundant operations

### ISSUE THREAD

**Issue by akutz** [kubernetes-sigs/cluster-api-provider-vsphere#684]
/kind bug

**What steps did you take and what happened:**
Using [a work-in-progress, feature branch](https://github.com/akutz/cluster-api-provider-vsphere/tree/feature/task-source) to address #655 I started to be able to reproduce an issue we've seen occasionally with CAPV -- a second clone / power / reconfigure operation for a VM. Ultimately everything works, but we've never been able to figure out why these redundant operations occur. The feature branch on which I'm working increases the number of reconcile requests for a given resource, and so the issue started occurring with more regularity. With additional logging I was able to narrow the issue to what looks like a problem patching the `VSphereMachine.Status` with the [`patch.Helper`](https://github.com/kubernetes-sigs/cluster-api/blob/65ee17c8f76f09496d457897a6c84100781ca24a/util/patch/patch.go#L30-L39).

**What did you expect to happen:**
A normal CAPV pivot should occur without any irregularities -- the status of a `VSphereMachine` patched to the API server should be equal to the status of the resource that was patched.

When the `VSphereMachine.Status.TaskRef` is incorrect coming into a reconcile request, a redundant operation, ex. clone, reconfigure, power on, will likely be the result. Users are worried about these failed operations occurring in the vSphere task collector. Why does CAPV try to clone the same VM twice? Why does CAPV try to power on a VM if it's already running? Unless the `VSphereMachine.Status.TaskRef` that is patched to the API server is reflected in subsequent reconcile requests, these redundant operations will continue to occur. If the issue is with the patch helper, then hopefully some of this logging output will assist in driving towards a solution.

The following code has been added to the `defer` block that is responsbile for patching the `VSphereMachine` with the patch helper:

```golang
// Always issue a patch when exiting this function so changes to the
// resource are patched back to the API server.
defer func() {
	// Patch the VSphereMachine resource.
	if err := machineContext.Patch(); err != nil {
		if reterr == nil {
			reterr = err
		}
		machineContext.Logger.Error(err, "patch failed", "machine", machineContext.String())
	}

	// localObj references the VSphereMachine resource fetched at the
	// beginning of this reconcile request.
	localObj := machineContext.VSphereMachine

	// remoteObj refererences the same VSphereMachine resource as it exists
	// on the API server post the patch operation above. In a perfect world,
	// the Status for localObj and remoteObj should be the same.
	var remoteObj *infrav1.VSphereMachine

	// Fetch the up-to-date VSphereMachine resource into remoteObj until the
	// fetched resource has a ResourceVersion equal to or greater than the
	// local resource.
	//
	// Since CAPV is currently deployed with a single replica, and this
	// controller has a max concurrency of one, the only agent updating the
	// VSphereMachine resource should be this controller.
	//
	// So if the remote resource's ResourceVersion is equal to or greater
	// than the ResourceVersion of the resource fetched at the beginning of
	// this reconcile request, then that means the remote resource *should*
	// be identical to the local resource.
	for {
		remoteObj = &infrav1.VSphereMachine{}
		if err := r.Client.Get(r, req.NamespacedName, remoteObj); err != nil {
			panic(errors.Wrap(err, "failed to get VSphereMachine from API server"))
		}
		rv1, _ := strconv.Atoi(localObj.ResourceVersion)
		rv2, _ := strconv.Atoi(remoteObj.ResourceVersion)
		if rv1 <= rv2 {
			machineContext.Logger.Info(
				"localResourceVersion lte remoteResourceVersion",
				"local-resource-version", rv1,
				"remote-resource-version", rv2)
			break
		}
		machineContext.Logger.Info(
			"localResourceVersion gt remoteResourceVersion",
			"local-resource-version", rv1,
			"remote-resource-version", rv2)
		time.Sleep(time.Second * 1)
	}
	if statusDiff := cmp.Diff(localObj.Status, remoteObj.Status); statusDiff != "" {
		buf := &bytes.Buffer{}
		fmt.Fprintf(buf, "\n\n")
		fmt.Fprintf(buf, "VSPHEREMACHINE STATUS SYNC ISSUE\n\n")
		fmt.Fprintf(buf, "STATUS DIFF\n\n%s\n\n", indent(statusDiff))
		resourceDiff := cmp.Diff(localObj, remoteObj)
		fmt.Fprintf(buf, "RESOURCE DIFF \n\n%s\n\n", indent(resourceDiff))
		io.Copy(os.Stdout, buf)
	}
}()
```

The additional logging has revealed a discrepancy between the local `VSphereMachine` resource's `Status` data and the resource patched back to the API server. For example, take the following excerpts from [capv-resource-drift.1.log](https://gist.github.com/akutz/f4b2f8a327ea78542842a374ab8e2c15#file-capv-resource-drift-1-log), such when [the difference is with the resource's `Status.TaskRef` field](https://gist.github.com/akutz/f4b2f8a327ea78542842a374ab8e2c15#file-capv-resource-drift-1-log-L96-L126):

```shell
VSPHEREMACHINE STATUS SYNC ISSUE

STATUS DIFF

      v1alpha2.VSphereMachineStatus{
      	Ready:        false,
      	Addresses:    nil,
    - 	TaskRef:      "task-275681",
    + 	TaskRef:      "task-275680",
      	Network:      nil,
      	ErrorReason:  nil,
      	ErrorMessage: nil,
      }
    

RESOURCE DIFF 

      &v1alpha2.VSphereMachine{
      	TypeMeta:   v1.TypeMeta{Kind: "VSphereMachine", APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha2"},
      	ObjectMeta: v1.ObjectMeta{Name: "sakc4-controlplane-0", Namespace: "default", SelfLink: "/apis/infrastructure.cluster.x-k8s.io/v1alpha2/namespaces/default/vspheremachines/sakc4-controlplane-0", UID: "f875ff30-8a9d-4f0b-95ea-d4cfb515e9ce", ResourceVersion: "728", Generation: 1, CreationTimestamp: v1.Time{Time: s"2019-12-08 06:29:09 +0000 UTC"}, Labels: map[string]string{"cluster.x-k8s.io/cluster-name": "sakc4", "cluster.x-k8s.io/control-plane": "true"}, OwnerReferences: []v1.OwnerReference{{APIVersion: "cluster.x-k8s.io/v1alpha2", Kind: "Machine", Name: "sakc4-controlplane-0", UID: "c8829c33-186d-4ccf-bcca-9b5df2f2042b"}}, Finalizers: []string{"vspheremachine.infrastructure.cluster.x-k8s.io"}},
      	Spec:       v1alpha2.VSphereMachineSpec{Template: "ubuntu-1804-kube-v1.16.2", Datacenter: "SDDC-Datacenter", Network: v1alpha2.NetworkSpec{Devices: []v1alpha2.NetworkDeviceSpec{{NetworkName: "sddc-cgw-network-5", DHCP4: true}}}, NumCPUs: 2, MemoryMiB: 2048, DiskGiB: 50},
      	Status: v1alpha2.VSphereMachineStatus{
      		Ready:        false,
      		Addresses:    nil,
    - 		TaskRef:      "task-275681",
    + 		TaskRef:      "task-275680",
      		Network:      nil,
      		ErrorReason:  nil,
      		ErrorMessage: nil,
      	},
      }
```

Or when [the difference includes part of the resource's `ObjectMeta`](https://gist.github.com/akutz/f4b2f8a327ea78542842a374ab8e2c15#file-capv-resource-drift-1-log-L221-L282) as well:

```shell
VSPHEREMACHINE STATUS SYNC ISSUE

STATUS DIFF

      v1alpha2.VSphereMachineStatus{
    - 	Ready:     true,
    + 	Ready:     false,
    - 	Addresses: []v1.NodeAddress{{Type: "InternalIP", Address: "192.168.5.209"}},
    + 	Addresses: nil,
      	TaskRef:   "",
      	Network: []v1alpha2.NetworkStatus{
      		{
    - 			Connected:   true,
    + 			Connected:   false,
    - 			IPAddrs:     []string{"192.168.5.209"},
    + 			IPAddrs:     nil,
      			MACAddr:     "00:50:56:b0:00:80",
      			NetworkName: "",
      		},
      	},
      	ErrorReason:  nil,
      	ErrorMessage: nil,
      }
    

RESOURCE DIFF 

      &v1alpha2.VSphereMachine{
      	TypeMeta: v1.TypeMeta{Kind: "VSphereMachine", APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha2"},
      	ObjectMeta: v1.ObjectMeta{
      		... // 3 identical fields
      		SelfLink:          "/apis/infrastructure.cluster.x-k8s.io/v1alpha2/namespaces/default/vspheremachines/sakc4-controlplane-0",
      		UID:               "f875ff30-8a9d-4f0b-95ea-d4cfb515e9ce",
    - 		ResourceVersion:   "792",
    + 		ResourceVersion:   "899",
    - 		Generation:        1,
    + 		Generation:        2,
      		CreationTimestamp: v1.Time{Time: s"2019-12-08 06:29:09 +0000 UTC"},
      		DeletionTimestamp: nil,
      		... // 8 identical fields
      	},
      	Spec: v1alpha2.VSphereMachineSpec{ProviderID: &"vsphere://42307788-cd8e-8610-49fc-f490b6529f04", Template: "ubuntu-1804-kube-v1.16.2", Datacenter: "SDDC-Datacenter", Network: v1alpha2.NetworkSpec{Devices: []v1alpha2.NetworkDeviceSpec{{NetworkName: "sddc-cgw-network-5", DHCP4: true}}}, NumCPUs: 2, MemoryMiB: 2048, DiskGiB: 50},
      	Status: v1alpha2.VSphereMachineStatus{
    - 		Ready:     true,
    + 		Ready:     false,
    - 		Addresses: []v1.NodeAddress{{Type: "InternalIP", Address: "192.168.5.209"}},
    + 		Addresses: nil,
      		TaskRef:   "",
      		Network: []v1alpha2.NetworkStatus{
      			{
    - 				Connected:   true,
    + 				Connected:   false,
    - 				IPAddrs:     []string{"192.168.5.209"},
    + 				IPAddrs:     nil,
      				MACAddr:     "00:50:56:b0:00:80",
      				NetworkName: "",
      			},
      		},
      		ErrorReason:  nil,
      		ErrorMessage: nil,
      	},
      }
```

In the first log the drift happened three times, but in another run, [`capv-resource-drift.2.log`](https://gist.github.com/akutz/f4b2f8a327ea78542842a374ab8e2c15#file-capv-resource-drift-2-log), the drift happened only once, albeit [with the `ObjectMeta` again](https://gist.github.com/akutz/f4b2f8a327ea78542842a374ab8e2c15#file-capv-resource-drift-2-log-L139-L201):

```shell
VSPHEREMACHINE STATUS SYNC ISSUE

STATUS DIFF

      v1alpha2.VSphereMachineStatus{
    - 	Ready:     true,
    + 	Ready:     false,
    - 	Addresses: []v1.NodeAddress{{Type: "InternalIP", Address: "192.168.5.106"}},
    + 	Addresses: nil,
      	TaskRef:   "",
      	Network: []v1alpha2.NetworkStatus{
      		{
    - 			Connected:   true,
    + 			Connected:   false,
    - 			IPAddrs:     []string{"192.168.5.106"},
    + 			IPAddrs:     nil,
      			MACAddr:     "00:50:56:b0:ab:0e",
      			NetworkName: "",
      		},
      	},
      	ErrorReason:  nil,
      	ErrorMessage: nil,
      }
    

RESOURCE DIFF 

      &v1alpha2.VSphereMachine{
      	TypeMeta: v1.TypeMeta{Kind: "VSphereMachine", APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha2"},
      	ObjectMeta: v1.ObjectMeta{
      		... // 3 identical fields
      		SelfLink:          "/apis/infrastructure.cluster.x-k8s.io/v1alpha2/namespaces/default/vspheremachines/sakc4-controlplane-0",
      		UID:               "7782f31c-5d5c-484a-ade0-aa7a73d354e1",
    - 		ResourceVersion:   "730",
    + 		ResourceVersion:   "833",
    - 		Generation:        1,
    + 		Generation:        2,
      		CreationTimestamp: v1.Time{Time: s"2019-12-08 06:55:20 +0000 UTC"},
      		DeletionTimestamp: nil,
      		... // 8 identical fields
      	},
      	Spec: v1alpha2.VSphereMachineSpec{ProviderID: &"vsphere://4230c21b-b3aa-a67c-ec81-86f1eded316a", Template: "ubuntu-1804-kube-v1.16.2", Datacenter: "SDDC-Datacenter", Network: v1alpha2.NetworkSpec{Devices: []v1alpha2.NetworkDeviceSpec{{NetworkName: "sddc-cgw-network-5", DHCP4: true}}}, NumCPUs: 2, MemoryMiB: 2048, DiskGiB: 50},
      	Status: v1alpha2.VSphereMachineStatus{
    - 		Ready:     true,
    + 		Ready:     false,
    - 		Addresses: []v1.NodeAddress{{Type: "InternalIP", Address: "192.168.5.106"}},
    + 		Addresses: nil,
      		TaskRef:   "",
      		Network: []v1alpha2.NetworkStatus{
      			{
    - 				Connected:   true,
    + 				Connected:   false,
    - 				IPAddrs:     []string{"192.168.5.106"},
    + 				IPAddrs:     nil,
      				MACAddr:     "00:50:56:b0:ab:0e",
      				NetworkName: "",
      			},
      		},
      		ErrorReason:  nil,
      		ErrorMessage: nil,
      	},
      }
```

**Anything else you would like to add:**
cc @vincepri 

**Environment:**

- Cluster-api-provider-vsphere version: https://github.com/akutz/cluster-api-provider-vsphere/tree/feature/task-source
- Kubernetes version: (use `kubectl version`):

  ```shell
  $ kubectl --kubeconfig ./out/sakc4/kubeconfig version
  Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:11:31Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"darwin/amd64"}
  Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:09:08Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
  ```

- OS (e.g. from `/etc/os-release`):

  ```shell
  $ cat /etc/os-release
  NAME="Ubuntu"
  VERSION="18.04.3 LTS (Bionic Beaver)"
  ID=ubuntu
  ID_LIKE=debian
  PRETTY_NAME="Ubuntu 18.04.3 LTS"
  VERSION_ID="18.04"
  HOME_URL="https://www.ubuntu.com/"
  SUPPORT_URL="https://help.ubuntu.com/"
  BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
  PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
  VERSION_CODENAME=bionic
  UBUNTU_CODENAME=bionic
  ```

**Comment by akutz** [comment]
And lo and behold the following change produced at least a single no-drift run at [`capv-resource-sync.1.log`](https://gist.github.com/akutz/f4b2f8a327ea78542842a374ab8e2c15#file-capv-resource-sync-1-log):

```golang
// Always issue a patch when exiting this function so changes to the
// resource are patched back to the API server.
defer func() {
	// Patch the VSphereMachine resource.
	if err := machineContext.Patch(); err != nil {
		if reterr == nil {
			reterr = err
		}
		machineContext.Logger.Error(err, "patch failed", "machine", machineContext.String())
	}

	// localObj references the VSphereMachine resource fetched at the
	// beginning of this reconcile request.
	localObj := machineContext.VSphereMachine

	// remoteObj refererences the same VSphereMachine resource as it exists
	// on the API server post the patch operation above. In a perfect world,
	// the Status for localObj and remoteObj should be the same.
	var remoteObj *infrav1.VSphereMachine

	// Fetch the up-to-date VSphereMachine resource into remoteObj until the
	// fetched resource has a ResourceVersion equal to or greater than the
	// local resource and both resource's have the same status.
	//
	// Since CAPV is currently deployed with a single replica, and this
	// controller has a max concurrency of one, the only agent updating the
	// VSphereMachine resource should be this controller.
	//
	// So if the remote resource's ResourceVersion is equal to or greater
	// than the ResourceVersion of the resource fetched at the beginning of
	// this reconcile request, then that means the remote resource's status
	//  *should* be identical to the local resource's status.
	for {
		remoteObj = &infrav1.VSphereMachine{}
		if err := r.Client.Get(r, req.NamespacedName, remoteObj); err != nil {
			panic(errors.Wrap(err, "failed to get VSphereMachine from API server"))
		}
		localResourceVersion, _ := strconv.Atoi(localObj.ResourceVersion)
		remoteResourceVersion, _ := strconv.Atoi(remoteObj.ResourceVersion)
		statusDiff := cmp.Diff(localObj.Status, remoteObj.Status)
		if localResourceVersion <= remoteResourceVersion && statusDiff == "" {
			machineContext.Logger.Info(
				"resource is patched",
				"local-resource-version", localResourceVersion,
				"remote-resource-version", remoteResourceVersion)
			break
		}
		machineContext.Logger.Info(
			"resource is not patched",
			"local-resource-version", localResourceVersion,
			"remote-resource-version", remoteResourceVersion)
		if statusDiff != "" {
			buf := &bytes.Buffer{}
			fmt.Fprintf(buf, "\n\n")
			fmt.Fprintf(buf, "VSPHEREMACHINE STATUS SYNC ISSUE\n\n")
			fmt.Fprintf(buf, "STATUS DIFF\n\n%s\n\n", indent(statusDiff))
			resourceDiff := cmp.Diff(localObj, remoteObj)
			fmt.Fprintf(buf, "RESOURCE DIFF \n\n%s\n\n", indent(resourceDiff))
			io.Copy(os.Stdout, buf)
		}
		time.Sleep(time.Second * 1)
	}
}()
```

**Comment by akutz** [comment]
cc @ncdc on the off-chance this is related to the patch helper...

**Comment by akutz** [comment]
And it looks like it **is** related to the patching. If not the helper, then the time it takes for the cache/API server to sync. The next run with the above change showed the drift again, but now we can see more interesting things about it:

```shell
I1208 07:32:24.360947       1 vspheremachine_controller.go:235] capv-controller-manager/vspheremachine-controller/default/sakc4-controlplane-0 "level"=0 "msg"="resource is not patched"  "local-resource-version"=639 "remote-resource-version"=639


VSPHEREMACHINE STATUS SYNC ISSUE

STATUS DIFF

      v1alpha2.VSphereMachineStatus{
      	Ready:        false,
      	Addresses:    nil,
    - 	TaskRef:      "task-275716",
    + 	TaskRef:      "",
      	Network:      nil,
      	ErrorReason:  nil,
      	ErrorMessage: nil,
      }
    

RESOURCE DIFF 

      &v1alpha2.VSphereMachine{
      	TypeMeta:   v1.TypeMeta{Kind: "VSphereMachine", APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha2"},
      	ObjectMeta: v1.ObjectMeta{Name: "sakc4-controlplane-0", Namespace: "default", SelfLink: "/apis/infrastructure.cluster.x-k8s.io/v1alpha2/namespaces/default/vspheremachines/sakc4-controlplane-0", UID: "f5dbd497-aaa6-4fb0-9cab-be109381dec9", ResourceVersion: "639", Generation: 1, CreationTimestamp: v1.Time{Time: s"2019-12-08 07:31:03 +0000 UTC"}, Labels: map[string]string{"cluster.x-k8s.io/cluster-name": "sakc4", "cluster.x-k8s.io/control-plane": "true"}, OwnerReferences: []v1.OwnerReference{{APIVersion: "cluster.x-k8s.io/v1alpha2", Kind: "Machine", Name: "sakc4-controlplane-0", UID: "5db61cd4-63dd-49e3-9da7-7cdd38aaea69"}}, Finalizers: []string{"vspheremachine.infrastructure.cluster.x-k8s.io"}},
      	Spec:       v1alpha2.VSphereMachineSpec{Template: "ubuntu-1804-kube-v1.16.2", Datacenter: "SDDC-Datacenter", Network: v1alpha2.NetworkSpec{Devices: []v1alpha2.NetworkDeviceSpec{{NetworkName: "sddc-cgw-network-5", DHCP4: true}}}, NumCPUs: 2, MemoryMiB: 2048, DiskGiB: 50},
      	Status: v1alpha2.VSphereMachineStatus{
      		Ready:        false,
      		Addresses:    nil,
    - 		TaskRef:      "task-275716",
    + 		TaskRef:      "",
      		Network:      nil,
      		ErrorReason:  nil,
      		ErrorMessage: nil,
      	},
      }
    

I1208 07:32:25.361676       1 vspheremachine_controller.go:229] capv-controller-manager/vspheremachine-controller/default/sakc4-controlplane-0 "level"=0 "msg"="resource is patched"  "local-resource-version"=639 "remote-resource-version"=727
I1208 07:32:25.361736       1 vspheremachine_controller.go:183] capv-controller-manager/vspheremachine-controller/default/sakc4-controlplane-0 "level"=4 "msg"="VSphereMachine.Status.TaskRef OnExit"  "task-ref"="task-275716"
```

The first log entry says `resource is not patched` and prints the diff between the resources. The only difference is the `VSphereMachine.Status.TaskRef`. Because the resource wasn't viewed yet as patched, the `for {}` loop continued to the next iteration and fetched the remote `VSphereMachine` resource again. This time the remote resource **did** have the same status as the local resource, and so we see `resource is patched`.

Please note that a second patch operation did **not** occur -- rather it seems like there is some amount of time that must be allowed before retrieving a remote resource accurately reflects the resource data that has been patched back to the API server.**

**Comment by akutz** [comment]
@ncdc posted [a nice tl;dr](https://kubernetes.slack.com/archives/CKFGK3SSD/p1575910983213500?thread_ts=1575906518.197400) of the underlying issue:

> tl;dr on the root cause: if you have multiple reconcile event sources, you have to deal with the possibility that your cache may be stale compared to what’s in etcd, and if your 2nd event source (in this case, a watch on vCenter) triggers a reconcile, you aren’t guaranteed that the object you get has the changes you made in the previous reconcile loop
